frontend/modules/tree/treeProcessing/processing.ts

```diff
@@ -0,0 +1,431 @@
+	}
+
+	public get selectedNodesIds() {
+		return keys(pickBy(this.selectionMap, (selectionState) => {
+			return selectionState !== SELECTION_STATES.UNSELECTED;
+		}));
+	}
+
+	public get highlightedNodesIds() {
+		return keys(pickBy(this.selectionMap, (selectionState) => {
+			return selectionState === SELECTION_STATES.SELECTED;
+		}));
+	}
+
+	public get invisibleNodesIds() {
+		return keys(pickBy(this.visibilityMap, (selectionState) => {
+			return selectionState === VISIBILITY_STATES.INVISIBLE;
+		}));
+	}
+
+	public clearCurrentlySelected = () => {
+		const selectedNodesIds = this.selectedNodesIds;
+		for (let index = 0, size = selectedNodesIds.length; index < size; index++) {
+			this.selectionMap[selectedNodesIds[index]] = SELECTION_STATES.UNSELECTED;
+		}
+
+		this.selectionMap = {...this.selectionMap};
+	}
+
+	public selectNodes = ({ nodesIds = [], ...extraData }) => {
+		const visibleNodesIds = nodesIds.filter((nodeId) => this.visibilityMap[nodeId] !== VISIBILITY_STATES.INVISIBLE);
+
+		if (!visibleNodesIds.length) {
+			return { highlightedObjects: [] };
+		}
+
+		const { skipChildren } = extraData;
+		let nodes = this.getNodesByIds(visibleNodesIds).filter((node) => {
+			return this.visibilityMap[node._id] !== VISIBILITY_STATES.INVISIBLE;
+		});
+
+		if (!skipChildren) {
+			const compactNodes = [];
+
+			for (let index = 0, size = nodes.length; index < size; index++) {
+				const node = nodes[index];
+				const children = this.getDeepChildren(node);
+				compactNodes.push(node, ...children);
+			}
+
+			nodes = compactNodes;
+		}
+
+		this.handleToSelect(nodes);
+		const highlightedObjects = this.getMeshesByNodes(nodes);
+
+		this.selectionMap = { ...this.selectionMap };
+		return { highlightedObjects };
+	}
+
+	public deselectNodes = ({ nodesIds = [] }) => {
+		const filteredNodesIds = intersection(nodesIds, this.selectedNodesIds);
+		const nodes = this.getNodesByIds(filteredNodesIds);
+
+		const children = nodes.map((node) => this.getDeepChildren(node)) as any;
+		const nodesWithChildren = [...nodes, ...children.flat()];
+
+		for (let index = 0; index < filteredNodesIds.length; index++) {
+			this.selectionMap[filteredNodesIds[index]] = SELECTION_STATES.UNSELECTED;
+		}
+
+		this.handleToDeselect(nodesWithChildren);
+
+		this.selectionMap = { ...this.selectionMap };
+
+		const unhighlightedObjects = this.getMeshesByNodes(nodesWithChildren);
+		return { unhighlightedObjects };
+	}
+
+	public isolateNodes = ({ nodesIds = [], ifcSpacesHidden = true }: any) => {
+		const toUnhighlight = [];
+		const toHighlight = [];
+		const meshesToUpdate = [];
+
+		for (let index = 0; index < this.nodesList.length; index++) {
+			const node = this.nodesList[index];
+			const shouldBeVisible = nodesIds.includes(node._id);
+			if (shouldBeVisible) {
+				if (ifcSpacesHidden) {
+					this.visibilityMap[node._id] = ifcSpacesHidden ? this.defaultVisibilityMap[node._id] : VISIBILITY_STATES.VISIBLE;
+				}
+
+				if (this.isVisibleNode(node._id)) {
```

Comment:
Assuming we dumped L113. Then we already know the state of the object because we set it.

So maybe we should set it to a variable and use it instead of calling a function to probe the value again?

Reply:
I added this function to make the code more "readable" - it contains a regular value check, that's all. It's also more convenient in case if you decide e.g to change the data structure.

Nevertheless, I refactored that part of code a bit.